server.go

```diff
@@ -346,6 +354,53 @@ func (server *mongoServer) pinger(loop bool) {
 	}
 }
 
+func (server *mongoServer) poolShrinker() {
+	ticker := time.NewTicker(1 * time.Minute)
+	for _ = range ticker.C {
+		if server.closed {
+			ticker.Stop()
+			return
+		}
+		server.Lock()
+		unused := len(server.unusedSockets)
+		if unused < server.minPoolSize {
+			server.Unlock()
+			continue
+		}
+		now := time.Now()
+		end := 0
+		reclaimMap := map[*mongoSocket]bool{}
+		// Because the acquirision and recycle are done at the tail of array,
+		// the head is always the oldest unused socket.
+		for _, s := range server.unusedSockets[:unused-server.minPoolSize] {
+			if s.lastTimeUsed.Add(time.Duration(server.maxIdleTimeMS) * time.Millisecond).After(now) {
+				break
+			}
+			end++
+			reclaimMap[s] = true
+		}
+		tbr := server.unusedSockets[:end]
+		if end > 0 {
+			next := make([]*mongoSocket, unused-end)
+			copy(next, server.unusedSockets[end:])
+			server.unusedSockets = next
+			remainSockets := []*mongoSocket{}
+			for _, s := range server.liveSockets {
```

Comment:
It is late, and I might be missing something here,  but it looks like you populate the `reclaimMap` with the unused sockets. Then you check if the `liveSockets` contain elements from the `unusedSockets`. Is it possible to `liveSockets`  and `unusedSockets` to contain pointers to the same structs ?

Reply:
If I understand correctly, the `liveSockets` includes both in-use sockets and `unusedSockets`, and when we reclaim the timeout elements in `unusedSockets`, they should be removed from `liveSockets` as well. If you don't remove them, there would lead to a "leakage", i.e. when you allocate new socket, there are full of live sockets but no unused, and no sockets could be allocated then. I observed the case during my local test.

Reply:
Right, thanks for the clarification :).